### ISSUE TITLE
Warping landmarks create out of bounds values

### ISSUE THREAD

**Issue by Machvrick** [menpo/menpo#806]
Hello,

First, thank you for providing us with this tool which has helped me a lot for my project. But I'm now facing an issue when trying to warp a video of a face into frontal position and warp also the landmarks.

I trained an AAM and warp each frame of the video into the mean shape of the AAM, this works fine and I get a ~115x110 video of the warped face. (I'm using ThinPlateSplines and not PiecewiseAffine because with the latter I get TrianglecontainmentError but that's another issue). 
But the problem is I also want to warp the landmarks to be able to extract a ROI around the mouth, so I pass the argument warp_landmarks=True in the call of the method warp_to_mask() on an image. But this produce really high value for the landmarks. For example, on the last frame the original position of the landmark is (106, 73) for a 720x480 video, and after warping (-1959, -2874) so out of the bounds.

Here is a piece of code to reproduce the issue, with the video, landmarks and trained AAM pickle:
https://drive.google.com/open?id=0B-89vzrA9ub3a0t3b3F2ZW5OQ0E

I had to create a function pad_to_even to pad an image so the height and width are even numbers, if not ffmpeg raises an error during the export.

**Comment by patricksnape** [comment]
So the issue is that when you warp your image to the reference frame you've already implicitly aligned the shapes together. This is possible because the non-rigid transformations such as `PiecewiseAffine` and `ThinPlateSplines` can align any set of points together 'perfectly'. This is why you were having issues with `PiecewiseAffine` as it was essentially telling you that the points are exactly on top of each other (hence the sampling error).

In short, you don't need to warp the landmarks because you already did. The landmarks are exactly where the source landmarks are in the reference frame. The movement you see in the video is actually due to the imperfect alignment between the frames.

If you want the non-rigid movement to be there you need to only align rigidly (use `AlignmentSimilarity`)

**Comment by Machvrick** [comment]
Thank you for the quick answer, I think I understand what the problem was but if I don't warp the landmarks with warp_landmarks=True, then the warped images don't have any landmarks. How can I access them if I want to continue to use `PiecewiseAffine` or `ThinPlateSplines` ?

I tried to use `AlignmentSimilarity` and it works well, the exported landmarks are correct. However I tried it on my other dataset which is in the wild (quite different from CUAVE from which the video I sent was) and I got the following error: 
`    raise OutOfMaskSampleError(sampled_mask, sampled_values)
menpo.image.masked.OutOfMaskSampleError
`

Do you have any idea from where this issue could come from ?
I have followed the same approach than on my first example: trained an AAM on 2000 annotated images then I take the mean shape and try to warp all my annotated videos to these mean shape in order to 'normalize' each image in the videos (because the head pose is different in each video and I need to have a sort of 'normalization')
I'm sorry but I cannot send this video because of the license agreement.

**Comment by patricksnape** [comment]
The warped images don't have landmarks, but the correct landmarks are just the source landmarks you use to warp in the first place. If you set the source landmarks as a landmark group on the warped images you will see that they align perfectly.

The `OutOfMaskSampleError` implies one of two things:

  1. You are warping `MaskedImage` into another `MaskedImage` - it is unlikely you want to do this. Try converting the input image to a normal Image - see `as_unmasked`
  2. The input shape has degenerate points - think two points directly on top of each other which causes instability in the transformation computation.

I think (1) is most likely - are you sure the images you warping are not masked?

**Comment by Machvrick** [comment]
Thank you very much, both issues are now solved. 

For the first problem you are right and I did not know that I have to consider the source landmarks.
For the second problem  `OutOfMaskSampleError`, you are also right and if I remove img.as_masked() in my code it works well and I have a correct warped image.
